### PR TITLE
Update Matrix/TableWorkspaceDisplay and InstrumentView on workspace change

### LINKED ISSUE #24475: Workbench: Update matrix/table data display on workspace changes
#24000 addresses updating plots when workspaces are changed/deleted but the same functionality needs to be added to the MatrixWorkspace and TableWorkspace data views.

This should follow a similar pattern as with the plots and use the `AnalysisDataServiceObserver` class to watch for updates.

### LINKED ISSUE #24499: Instrument view doesn't disappear with delete workspace
Minor bug:
1. Load data
2. Instrument view it
3. Delete the workspace with instrument view still up
4. See that the contents of the window are gone, but the overall window is still there

This is found after merging #24414 into `master`.

### PR DESCRIPTION
**Pull request by dtasev**
**Description of work.**
Adds ADS observers and handles replace/rename/delete/clear events for the Matrix/Table workspace displays, and the instrument view

**To test:**
- Matrix Workspace
  - Load a matrix workspace
  - Show its data
  - Multiply the workspace and see that the numbers are updated
  - Delete the workspace and see that the window is closed
- Table Workspace
  - Do the same workflow as for Matrix
  - A good algorithm to modify the workspace is `SortTableWorkspace` as it will work with any table
- Open the instrument view, and delete the workspace, the window should be closed. This PR does **not** handle replace/rename events for instrument view, as they are handled in C++ `InstrumentWidget`

Fixes #24475 
Fixes #24499

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/python/mantidqt/widgets/common/test/test_observing_presenter.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
+# Mantid Repository : https://github.com/mantidproject/mantid
+#
+# Copyright &copy; 2019 ISIS Rutherford Appleton Laboratory UKRI,
+#     NScD Oak Ridge National Laboratory, European Spallation Source
+#     & Institut Laue - Langevin
+# SPDX - License - Identifier: GPL - 3.0 +
+#  This file is part of the mantid workbench.
+
```
Comment (martyngigg) [inline qt/python/mantidqt/widgets/common/test/test_observing_presenter.py]: Unnecessary whitespace I think?

File: qt/python/mantidqt/widgets/common/observing_view.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
+# Mantid Repository : https://github.com/mantidproject/mantid
+#
+# Copyright &copy; 2019 ISIS Rutherford Appleton Laboratory UKRI,
+#     NScD Oak Ridge National Laboratory, European Spallation Source
+#     & Institut Laue - Langevin
+# SPDX - License - Identifier: GPL - 3.0 +
+#  This file is part of the mantid workbench.
+
```
Comment (martyngigg) [inline qt/python/mantidqt/widgets/common/observing_view.py]: `__future__` imports?

File: qt/python/mantidqt/widgets/common/observing_presenter.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
+# Mantid Repository : https://github.com/mantidproject/mantid
+#
+# Copyright &copy; 2018 ISIS Rutherford Appleton Laboratory UKRI,
+#     NScD Oak Ridge National Laboratory, European Spallation Source
+#     & Institut Laue - Langevin
+# SPDX - License - Identifier: GPL - 3.0 +
+#  This file is part of the mantid workbench.
+
+
+class ObservingPresenter:
```
Comment (martyngigg) [inline qt/python/mantidqt/widgets/common/observing_presenter.py]: Can we use a new-style class while we're still in Python 2?

File: qt/python/mantidqt/widgets/common/observing_view.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+class ObservingView:
```
Comment (martyngigg) [inline qt/python/mantidqt/widgets/common/observing_view.py]: Can we use a new-style class while we're still in Python 2?

File: qt/python/mantidqt/widgets/common/observing_view.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    through inheritance.
+
+    This class shouldn't inherit from a QObject/QWidget, otherwise the closeEvent
+    doesn't replace the closeEvent of the view that is inheriting this. This also
+    makes it easily testable, as the GUI library isn't pulled in.
+    """
+
+    TITLE_STRING = "{} - Mantid"
+
+    def __init__(self, _):
```
Comment (martyngigg) [inline qt/python/mantidqt/widgets/common/observing_view.py]: It's not immediately obvious to why we need the `__init__`? I assume I've missed something so can we add a comment?

File: qt/python/mantidqt/widgets/common/workspacedisplay_ads_observer.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
+# Mantid Repository : https://github.com/mantidproject/mantid
+#
+# Copyright &copy; 2017 ISIS Rutherford Appleton Laboratory UKRI,
+#     NScD Oak Ridge National Laboratory, European Spallation Source
+#     & Institut Laue - Langevin
+# SPDX - License - Identifier: GPL - 3.0 +
+#  This file is part of the mantid workbench.
+#
+#
+
```
Comment (martyngigg) [inline qt/python/mantidqt/widgets/common/workspacedisplay_ads_observer.py]: `__future__` imports...

File: qt/python/mantidqt/widgets/instrumentview/presenter.py [hunk 7]
```diff
@@ -13,15 +13,52 @@
...
     Presenter holding the view widget for the InstrumentView.
     It has no model as its an old widget written in C++ with out MVP
     """
+
     view = None
 
-    def __init__(self, ws, parent=None):
-        self.view = InstrumentView(self, ws.name(), parent)
+    def __init__(self, ws, parent=None, ads_observer=None):
+        self.ws_name = ws.name()
```
Comment (martyngigg) [inline qt/python/mantidqt/widgets/instrumentview/presenter.py]: Do we not need to call `__init__`?

**Review by martyngigg — CHANGES_REQUESTED** [review]
Quick review so far...

Looks good. A few questions.

File: qt/python/mantidqt/widgets/common/observing_presenter.py (hunk 3 above)
Comment (DTasev) [reply qt/python/mantidqt/widgets/common/observing_presenter.py]: I've been running this on Python 2.7 and have had no issues

File: qt/python/mantidqt/widgets/common/observing_view.py (hunk 5 above)
Comment (DTasev) [reply qt/python/mantidqt/widgets/common/observing_view.py]: I thought I needed it because of the view's `super(..., self).__init__(parent)` call, which calls both the Qt constructor, and then this one, but it turns out there's no RuntimeError if I remove it. 

I would prefer to leave it in though, as it removes a `Unresolved attribute reference '__init__' for class 'ObservingView'` warning in PyCharm, which might be confusing in its own right.

File: qt/python/mantidqt/widgets/instrumentview/presenter.py (hunk 7 above)
Comment (DTasev) [reply qt/python/mantidqt/widgets/instrumentview/presenter.py]: The `ObservingPresenter`'s __init__ is empty

File: qt/python/mantidqt/widgets/common/observing_view.py (hunk 5 above)
Comment (martyngigg) [reply qt/python/mantidqt/widgets/common/observing_view.py]: That's a fair reason. :+1: for the comment

**Review by martyngigg — DISMISSED** [review]

**Comment by DTasev** [comment]
Fix has been pushed, it's ready for a quick review of the functionality to edit cells. Have a look at the `setCell` overload code, I am not sure if this is the best way to do it. It keeps the old behaviour by default so it shouldn't break any existing secCell code.

File: Framework/PythonInterface/mantid/api/src/Exports/ITableWorkspace.cpp [hunk 8]
```diff
@@ -677,6 +692,14 @@ void export_ITableWorkspace() {
            "number then it is interpreted as a row otherwise it "
            "is interpreted as a column name.")
 
+      .def("setCell", &setCellNoNotify,
```
Comment (martyngigg) [inline Framework/PythonInterface/mantid/api/src/Exports/ITableWorkspace.cpp]: This will have changed the call signature on the Python side so that it requires the extra argument. We can use `BOOST_PYTHON_FUNCTION_OVERLOADS` to provide a default argument. Have a look at the [`WorkspaceFactory`](https://github.com/mantidproject/mantid/blob/master/Framework/PythonInterface/mantid/api/src/Exports/WorkspaceFactory.cpp#L60) export

File: qt/python/mantidqt/widgets/common/test/test_workspacedisplay_ads_observer.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
+# Mantid Repository : https://github.com/mantidproject/mantid
+#
+# Copyright &copy; 2019 ISIS Rutherford Appleton Laboratory UKRI,
+#     NScD Oak Ridge National Laboratory, European Spallation Source
+#     & Institut Laue - Langevin
+# SPDX - License - Identifier: GPL - 3.0 +
+#  This file is part of the mantid workbench.
+
```
Comment (martyngigg) [inline qt/python/mantidqt/widgets/common/test/test_workspacedisplay_ads_observer.py]: Unnecessary whitespace. Can we have `__future__` directly under the comment?

File: qt/python/mantidqt/widgets/common/test_mocks/mock_mantid.py [hunk 10]
```diff
@@ -5,44 +5,16 @@
 #     & Institut Laue - Langevin
 # SPDX - License - Identifier: GPL - 3.0 +
 #  This file is part of the mantid workbench.
-#
-#
+
```
Comment (martyngigg) [inline qt/python/mantidqt/widgets/common/test_mocks/mock_mantid.py]: Remove whitespace?

File: qt/python/mantidqt/widgets/common/test_mocks/mock_qt.py [hunk 11]
```diff
@@ -1,12 +1,14 @@
...
 #
-# Copyright &copy; 2018 ISIS Rutherford Appleton Laboratory UKRI,
+# Copyright &copy; 2019 ISIS Rutherford Appleton Laboratory UKRI,
 #     NScD Oak Ridge National Laboratory, European Spallation Source
 #     & Institut Laue - Langevin
 # SPDX - License - Identifier: GPL - 3.0 +
 #  This file is part of the mantid workbench.
-#
-#
+
```
Comment (martyngigg) [inline qt/python/mantidqt/widgets/common/test_mocks/mock_qt.py]: Unnecessary whitespace?

File: qt/python/mantidqt/widgets/instrumentview/presenter.py (hunk 7 above)
Comment (martyngigg) [reply qt/python/mantidqt/widgets/instrumentview/presenter.py]: Should the `InstrumentViewPresenter` necessarily know that? It seems like good practice to make sure the base class inits are called?

**Review by martyngigg — CHANGES_REQUESTED** [review]

File: Framework/PythonInterface/mantid/api/src/Exports/ITableWorkspace.cpp [hunk 12]
```diff
@@ -687,14 +676,7 @@ void export_ITableWorkspace() {
...
            (arg("self"), arg("row_or_column"), arg("column_or_row"),
-            arg("value")),
-           "Sets the value of a given cell. If the row_or_column argument is a "
-           "number then it is interpreted as a row otherwise it "
-           "is interpreted as a column name.")
-
-      .def("setCell", &setCellNoNotify,
-           (arg("self"), arg("row_or_column"), arg("column_or_row"),
-            arg("value"), arg("notify_replace")),
+            arg("value"), arg("notify_replace") = true),
```
Comment (martyngigg) [inline Framework/PythonInterface/mantid/api/src/Exports/ITableWorkspace.cpp]: :+1:

**Review by martyngigg — APPROVED** [review]
Good to go now.
